src/python/py/models/builder.py

```diff
@@ -2613,6 +2642,48 @@ def make_layer(self, layer_id, layer):
             self.layernorm_attrs["last_layernorm"] = True
 
 
+class ChatGLMModel(Model):
+    def __init__(self, config, io_dtype, onnx_dtype, ep, cache_dir, extra_options):
+        super().__init__(config, io_dtype, onnx_dtype, ep, cache_dir, extra_options)
+        # self.input_shapes["position_ids"] = [1]  # Note: This is optional and only needed if you want position_ids to be an int instead of a 2D tensor
+        self.layernorm_attrs["simple"] = True # RMSNorm in ChatGLM
+        self.rotemb_attrs["num_heads"] = self.num_attn_heads
+        self.rotemb_attrs["partial_rotary_factor"] = 0.5 # Line 755 of modeling_chatglm.py check self.rotary_pos_emb declaration
+        self.rotemb_attrs["rotary_embedding_dim"] = int(self.head_size * self.rotemb_attrs["partial_rotary_factor"])
+        self.rotemb_attrs["interleaved"] = 1
+        self.mlp_attrs["use_proj"], self.mlp_attrs["use_fc"] = True, False
+        self.attention_attrs["use_rotemb_in_attn"] = True
+        self.attention_attrs["use_packed_matmul"] = True
+        self.attention_attrs["op_type"] = "GroupQueryAttention" if self.multi_query_attention else self.attention_attrs["op_type"]
+
+    def has_final_norm(self, module, model):
+        # Hugging Face names
+        hf_norm = hasattr(model, "model") and hasattr(model.model, "norm") and module == model.model.norm
```

Comment:
Can the changes in `ChatGLMModel.has_final_norm` be upstreamed to `Model.has_final_norm`? Then, this method doesn't have to be re-defined.

Reply:
Should we change `Model.has_final_norm` like this?

```python
def has_final_norm(self, module, model):
        # Hugging Face names
        hf_norm = hasattr(model, "model") and hasattr(model.model, "norm") and module == model.model.norm
        
        if(hasattr(model, "transformer")):
            hf_final_layernorm = hasattr(model.transformer, "encoder") and hasattr(model.transformer.encoder, "final_layernorm") and module == model.transformer.encoder.final_layernorm
        else:
            hf_final_layernorm = hasattr(model.model, "final_layernorm") and module == model.model.final_layernorm
        
        # GGUF names
        gguf_final_norm = hasattr(model, "final_norm") and module == model.final_norm
        return hf_norm or hf_final_layernorm or gguf_final_norm

Reply:
I think you can just create another variable for ChatGLM's logic and add that variable to the `return` logic.

```python
def has_final_norm(self, module, model):
    # Hugging Face names
    hf_norm = hasattr(model, "model") and hasattr(model.model, "norm") and module == model.model.norm
    hf_final_layernorm = hasattr(model, "model") and hasattr(model.model, "final_layernorm") and module == model.model.final_layernorm
    hf_transformer_final_layernorm = hasattr(model, "transformer") and hasattr(model.transformer, "encoder") and hasattr(model.transformer.encoder, "final_layernorm") and module == model.transformer.encoder.final_layernorm
    # GGUF names
    gguf_final_norm = hasattr(model, "final_norm") and module == model.final_norm
    return hf_norm or hf_final_layernorm or hf_transformer_final_layernorm or gguf_final_norm
```